### ISSUE TITLE
specialize_branched_loops memory bug

### ISSUE THREAD

**Issue by abadams** [halide/Halide#645]
@kree-colemcalughlin , The following program crashes with an out-of-bounds write. Disabling specialize_branched_loops fixes it. 

    #include <Halide.h>
    #include <stdio.h>

    using namespace Halide;
    
    int main(int argc, char **argv) {
        Var x, y;
        Func f, g;
        
        f(x, y) = x + y;
        
        g(x, y) = f(min(x, y), max(x, y));
        f.compute_at(g, y);
        g.vectorize(x, 16);
        
        Image<int> out = g.realize(16, 16);
        
        return 0;
    }

**Comment by kree-colemcalughlin** [comment]
I'll take a look at it.

**Comment by abadams** [comment]
Thanks.

**Comment by kree-colemcalughlin** [comment]
This bug is pretty gnarly. I am still working on it, but here are a few comments if you are interested:

1) Running the program with HL_TRACE=1 doesn't turn up any out-of-bounds stores, which is what I assume the problem is (indeed valgrind shows there are invalid reads/writes), and further more the program runs to completion without an error in this mode.

2) If you fully bound the output buffer, e.g.:
```
    g.output_buffer()
        .set_bounds(0, 0, 16)
        .set_bounds(1, 0, 16);
```
then the error goes away. I know this is not a solution, but could be a work around for now while I investigate further.

3) The fundamental issue seems to be difficulty in simplifying min/max exprs. In this particular case specialize_clamped_ramps generates a predicate that is tautologically false,
```
if ((((f1.s0.v0.v6.base + 15) <= f1.s0.v1) && (f1.s0.v1 <= f1.s0.v0.v6.base)))
```
but the simplifier can't prove that this condition is false, nor can any of the tools I built for specialize_branched_loops. Ultimately, in this particular case specialize_branched_loops is being triggered when it shouldn't be, for this if stmt is meaningless. However, it is still troubling that the computed branch boundaries do not seem to be valid in this case, and seem to indicate that somewhere along the way an invalid min or extent is getting set. 

tl;dr stay tuned...

**Comment by steven-johnson** [comment]
Obsolete?
